src/type-layout.md

```diff
@@ -285,40 +286,223 @@ the default `enum` size and alignment for the target platform's C ABI.
+using a field-less enum in FFI to model a C `enum` is often wrong.
 
 </div>
 
-It is an error for [zero-variant enumerations] to have the `C` representation.
+#### \#[repr(C)] Enums With Fields
 
-For all other enumerations, the layout is unspecified.
+The representation of a `repr(C)` enum with fields is a `repr(C)` struct with
+two fields, also called a "tagged union" in C:
 
-Likewise, combining the `C` representation with a primitive representation, the
-layout is unspecified.
+- a `repr(C)` version of the enum with all fields removed ("the tag")
+- a `repr(C)` union of `repr(C)` structs for the fields of each variant that had
+  them ("the payload")
+
+> Note: Due to the representation of `repr(C)` structs and unions, if a variant
+> has a single field there is no difference between putting that field directly
+> in the union or wrapping it in a struct; any system which wishes to manipulate
+> such an `enum`'s representation may therefore use whichever form is more
+> convenient or consistent for them
+
+```rust
+// This Enum has the same layout as ...
+#[repr(C)]
+enum MyEnum {
+    A(u32),
+    B(f32, u64),
+    C { x: u32, y: u8 },
+    D,
+ }
+
+// ... this struct.
+#[repr(C)]
+struct MyEnumRepr {
+    tag: MyEnumDiscriminant,
+    payload: MyEnumFields,
+}
+
+// This is the discriminant enum.
+#[repr(C)]
+enum MyEnumDiscriminant { A, B, C, D }
+
+// This is the variant union.
+#[repr(C)]
+union MyEnumFields {
+    A: MyAFields,
+    B: MyBFields,
+    C: MyCFields,
+    D: MyDFields,
+}
+
+#[repr(C)]
+#[derive(Copy, Clone)]
+struct MyAFields(u32);
+
+#[repr(C)]
+#[derive(Copy, Clone)]
+struct MyBFields(f32, u64);
+
+#[repr(C)]
+#[derive(Copy, Clone)]
+struct MyCFields { x: u32, y: u8 }
+
+// This struct could be omitted (it is a zero-sized type), and it must be in
+// C/C++ headers.
+#[repr(C)]
+#[derive(Copy, Clone)]
+struct MyDFields;
+```
+
+> Note: `union`s with non-`Copy` fields are unstable, see [55149].
 
 ### Primitive representations
 
 The *primitive representations* are the representations with the same names as
 the primitive integer types. That is: `u8`, `u16`, `u32`, `u64`, `u128`,
 `usize`, `i8`, `i16`, `i32`, `i64`, `i128`, and `isize`.
 
-Primitive representations can only be applied to enumerations.
+Primitive representations can only be applied to enumerations and have
+different behavior whether the enum has fields or no fields. It is an error
+for [zero-variant enumerations] to have a primitive representation. Combining
+two primitive representations together is an error.
+
+> Note: Primitive representations are primarily intended for Rust code that
+> wants to interoperate with the idioms of preexisting C/C++ codebases. If you
+> have control over both the Rust and C code, such as using C as FFI glue
+> between Rust and some third language, then you should use a
+> [primitive representation](#primitive-representation-of-enums-with-fields)
+> instead.
```

Comment:
I'm a little unclear here.  "instead" of what?  `repr(C)`?

Reply:
I'm not sure this whole note is useful. I could shorten it to:

> Note: Primitive representations are primarily intended for Rust code that wants to interoperate with the idioms of preexisting C/C++ codebases.

Reply:
I used the above wording in the "Fix nits 2" commit

Reply:
Reading it again, I'm a little unclear, since there are multiple forms.  In this case, is it referring to `repr(C, Int)` or `repr(Int)`?  AIUI:

* `repr(C, Int)` is useful for C/C++ code that uses a struct with an explicitly-sized tag and payload (which is probably more common)
* `repr(Int)` is useful for C/C++ code that uses a union with an explicitly-sized tag in each struct, and can be more space-efficient than `repr(C, Int), but less common for pre-existing C code?
* `repr(C)` is useful for C/C++ code that uses a struct with a **C-enum tag** and payload? And that tag size is implementation-defined, and "best guess", so probably harder to get correct? That is, you really need to know which C compiler, platform, and C flags are being used?

Is that roughly correct?

Reply:
I would say you are correct. I don't have much experience with this but the hobby projects I did used `repr(C, int)` to ensure maximal compatibility, which coincides with what you said. Would you like me to properly write in the document ?

Reply:
Hm.  How about removing the note altogether for now?  This feels like it would be more useful in a guide on using FFI, and I don't want to dive too deep on it here. The reference has generally shied away from including information on how something is supposed to be used, as that tends more toward guide material.

Reply:
I removed it.